### ISSUE TITLE
Could not push update of a csv resource to the datastore

### ISSUE THREAD

**Issue by Mbrownshoes** [bcgov/ckanext-bcgov#285]
After uploading a new csv to the filestore, the file was not able to push to the datastore. The following error was displayed in the datapusher window:

ERROR: Error getting datastore resource

I was able to upload the same csv file to a new resource. This is the file that used to powers the broker. 

After creating the new resource I deleted the old one.

CITZEDC-887

**Comment by ll911** [comment]
sounds like resource been used creates read lock in pg, so resource cannot be changed until lock removed.

**Comment by Mbrownshoes** [comment]
@ll911 is this a result of the broker pinging it? Is there a way to prevent it from locking up? We need this resource to be able to update smoothly, without having to recreate a resource and redeploy the broker.

**Comment by ll911** [comment]
workdaround/fix: delete resource view and repush to datastore

**Comment by kfishwick** [comment]
Has this occurred in cad or cat?  Are there logs in production that would indicate a read lock in pg (sounds like more of a "write" lock to me if you can't edit it).

@ll911  did you check postgres to see if there was indeed a locking error?

**Comment by Mbrownshoes** [comment]
@kfishwick we have not yet been able to reproduce in cat or cad. I've tried using both admin and editor accounts. There's only been two records in Prod that have experienced this, both in the past 2 days and both had resources with apis that are being consumed, one is used for the broker, and the second is used in a visualization on data.gov

**Comment by Mbrownshoes** [comment]
@kfishwick we have another resource that is locked and @ll911 's trick for deleting the resource view isn't working. These are popular resources that are being hit via their api. This is a big problem.  We need to determine what is causing these resources to lock up.
@gjlawran @dkelsey

**Comment by kfishwick** [comment]
@Mbrownshoes  what is the current resource that is locked?

**Comment by kfishwick** [comment]
@jrods  can you please check logs etc & see if you can find out anything?

**Comment by jrods** [comment]
@Mbrownshoes @ll911 have either of you done a datapusher_status api request for that resource? that api call looks like it can return an error message for the resource job, my api key is not authorized to make the request on prod.

**Comment by jrods** [comment]
looks like that api call is undocumented:
```
POST /api/action/datapusher_status HTTP/1.1
Host: catalogue.data.gov.bc.ca
Authorization: <your catalogue api key>
Cache-Control: no-cache
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW

------WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="id"

de425d68-a87f-44ce-ab4d-fcb514c04975
------WebKitFormBoundary7MA4YWxkTrZu0gW--
```

**Comment by jrods** [comment]
@Mbrownshoes can you provide detailed steps to how you were getting the initial error?

**Comment by Mbrownshoes** [comment]
@jrods @ll911 both resources where this issue has been observed are being accessed via their datastore apis, one is being used to create a visualization and the other is being used by the broker. We encountered this error when trying to manually update the resource. We simply managed the resource, removed the current file in the filestore and uploaded the new file. After noticing the data explorer and api button were not loading, we checked the datapusher where we encountered the error above.

**Comment by Mbrownshoes** [comment]
@jrods any progress on this? We will be having some data refreshes happening soon.

**Comment by Mbrownshoes** [comment]
We just experienced this again and had an important resource lock up.

**Comment by jrods** [comment]
@Mbrownshoes I'll have a look at the logs again. Still haven't been able to replicate in my own local env yet.

**Comment by ll911** [comment]
same pacakge, resource get's locked when datapusher trying to truncate the old resource when updates added via UI

**Comment by ll911** [comment]
the lock stuck for over 2 hours after one of the resource updated.

**Comment by Mbrownshoes** [comment]
This is only happening with resource that have datastore apis being consumed.

**Comment by Khalegh-H3** [comment]
This could happen if the csv file that is updated is being used by another consumer. It is the same as if you want to delete a file in your computer while it is still open in another program.

Uploading a new csv file means removing the old file from datastore and create a new one. That's why it always works fine for a new resource but it could cause issues for the available resources.

**Comment by Mbrownshoes** [comment]
@Khalegh-H3 we've tested other resources who's datastore apis are being hit and have not observed this problem. One of these resource apis is being used to power a visualization and the other a simple webpage, so I'm not sure if anything is being left 'open'. Any thoughts @ll911 ?

**Comment by gjlawran** [comment]
What attempts have already been made to replicate the problem? Do we have a Python script that makes the same calls as the Broker DataStore search and the DataStore update API calls?  

@Mbrownshoes  can you  please confirm that we have locking occur both through API update and when update is attempted via web interface?

**Comment by Mbrownshoes** [comment]
A resource_update did not cause locking, however we made that call on a resource that was already locked up. That resource then because unaccessible until there was a catalogue restart.

**Comment by ll911** [comment]
@Mbrownshoes resource is unaccessible when access exclusive lock is on, restart ckan or remove pg lock both can bring the access back, however at the moment we are unclear what process keep the resource "open" when editor update another resource within the same package

**Comment by Mbrownshoes** [comment]
@ll911 yesterday we were able to update the resource that was getting locked up previous weeks. Not sure why we were able to push successfully this time.

**Comment by Mbrownshoes** [comment]
@jrods @ll911 @gjlawran Once again we have a resource that won't push to the datastore. @kfishwick I'm letting you know now, as requested, before we implement our work around of creating a new resource.

**Comment by dkelsey** [comment]
**We feel we have a resolution to this. **

_NOTE: this is an edited version of the update email I sent out_

Mat added the mental-health resource to a record in CAT.
He also updated the broker (in PROD) to utilized this resource in CAT.
The motivation was that "we've never experienced the locking issue in CAT"
it turns out CAT has more verbose logging configured than PROD and this proved useful.

There was another issue in CAT at the time: I was looking at the logs and it seemed something was requesting every resource and had been doing so for as far back as the logs went.

There was another issue this week where CAT was failing: backup, stats and restart jobs were failing since the 19th.

In diagnosing the problem I found many many request for the ***same*** resource in CAT - at this point in time I didn't know that Mat had switched the resource as described above.

After Steve got CAT up and running again we learned that all request to CAT are passing through a reverse proxy on `canopus`.  The storm of requests  were for the `mental-health` resource.

Steve, Leo, Mat and I discussed and resolved the origin of the storm of these requests: 

https://cat.data.gov.bc.ca/api/action/datastore_search?resource_id=3edf2c73-9ed0-42c4-81b0-24871aab08e3&offset=0&limit=100

**It's the broker (running in OpenShift).**

OpenShift hits the `broker:8080/`   regularly (once a second?) - this is necessary: OpenShift is asserting that the container is running and the service endpoint is up and healthy.
The broker handles the request by performing a GET against the CKAN datastore for the mental-health resource.  It was pulling it from the datastore and re-indexing it for each health check!

Why is this happening in CAT?  (I'm repeating myself here but) Mat configured the broker to hit CAT because we never experienced the resource locking issue there.
Mat got the resource locking issue Wednesday afternoon on the resource in CAT (after we had figure out that it was the broker causing the lock).

I think we should confirm that we can reproduce the issue to be certain we got it: lets do this in CAT.

**Some ideas on how to resolve this were discussed:**

- ​if the broker needs to be on at all times, its response to the health check should not lock the resource in CAT.  The intention with the broker was for it to react instantly when the resource is updated.
- If the broker only needs to be on for a period after the resource is updated we discussed (Leo's suggestion) ​that we orchestrate a pipeline in Jenkins that is triggered when the resource is updated.

Leo stated that OpenShift has an API - We could turn the broker off after a couple of days or provide an API to GSA so that they turn if off when they are done.

**The current process is heavy touch:**
- Mat has to clean the CSV, 
- generate a KML, 
- start the broker, and 
- let the GSA people know the broker has been updated.
This can be automated, documented and instrumented.

**Where this is now:**

I've modified the broker adding a `/health` endpoint which simply returns `http 200`

```
curl -i http://localhost:8080/health
HTTP/1.1 200 OK
Content-Type: text/html
Date: Thu, 25 May 2017 18:22:25 GMT
Connection: keep-alive
Transfer-Encoding: chunked

<!DOCTYPE html><html><head><body><h1>ok</h1></body>
```

The change is in GOGS

**Comment by dkelsey** [comment]
I fixed a bug that would crash the broker if `:8080/search.html` were requested.

**Comment by dkelsey** [comment]
This was an issue triggered by the broker.  The fix was completed (adding a `/health` endpoint that simply returned `http 200`)
This drastically reduced the potential of the resource lock issue but does not eliminate it.  I'll suggest that the root of the problem is that reading from the datastore locks the resource and uploading (deleting and adding) does not check for a lock and blocks locking the resource.  I'm not completely satisfied: I wanted to distinguish between "we've reduced the probability of the locking issue occurring" and "we have eliminated any probability of the locking issue occurring"
